internal/restorer/restorer_windows_test.go

```diff
@@ -33,3 +41,621 @@ func getBlockCount(t *testing.T, filename string) int64 {
+			if existingFileAttr.Encrypted {
+				err := createEncryptedFileWriteData(filepath, nodeInfo)
+				rtest.OK(t, err)
+			} else {
+				// Write the data to the file
+				file, err := os.OpenFile(path.Clean(filepath), os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0600)
+				rtest.OK(t, err)
+				_, err = file.Write([]byte(nodeInfo.data))
+				rtest.OK(t, err)
+
+				err = file.Close()
+				rtest.OK(t, err)
+			}
+		} else {
+			os.MkdirAll(path.Join(testDir, nodeInfo.parentDir, nodeInfo.name), os.ModeDir)
+		}
+	}
+
+	if nodeInfo.Exists {
+		pathPointer, err := syscall.UTF16PtrFromString(path.Join(testDir, nodeInfo.parentDir, nodeInfo.name))
+		rtest.OK(t, err)
+		syscall.SetFileAttributes(pathPointer, GetAttributeValue(&existingFileAttr))
+	}
+
+	index := 0
+	order := []int{}
+
+	if !nodeInfo.IsDirectory {
+		order = append(order, index)
+		index++
+	}
+
+	streams := []DataStreamInfo{}
+	if !nodeInfo.IsDirectory {
+		streams = append(streams, nodeInfo.DataStreamInfo)
+	}
+	return setup(t, getNodes(nodeInfo.parentDir, nodeInfo.name, order, streams, nodeInfo.IsDirectory, &nodeInfo.attributes)), order
+}
+
+func getNodes(dir string, mainNodeName string, order []int, streams []DataStreamInfo, isDirectory bool, attributes *Attributes) []NamedNode {
+	var mode os.FileMode
+	if isDirectory {
+		mode = os.FileMode(2147484159)
+	} else {
+		if attributes != nil && attributes.ReadOnly {
+			mode = os.FileMode(292)
+		} else {
+			mode = os.FileMode(438)
+		}
+	}
+
+	getFileNodes := func() []NamedNode {
+		nodes := []NamedNode{}
+		if isDirectory {
+			//Add a directory node at the same level as the other streams
+			nodes = append(nodes, NamedNode{
+				name: mainNodeName,
+				node: OrderedDir{
+					ModTime:    time.Now(),
+					attributes: attributes,
+					Mode:       mode,
+				},
+				attributes: attributes,
+			})
+		}
+
+		if len(streams) > 0 {
+			for _, index := range order {
+				stream := streams[index]
+
+				var attr *Attributes = nil
+				if mainNodeName == stream.name {
+					attr = attributes
+				} else if attributes != nil && attributes.Encrypted {
+					//Set encrypted attribute
+					attr = &Attributes{Encrypted: true}
+				}
+
+				nodes = append(nodes, NamedNode{
+					name: stream.name,
+					node: File{
+						ModTime: time.Now(),
+						Data:    stream.data,
+						Mode:    mode,
+					},
+					attributes: attr,
+				})
+			}
+		}
+		return nodes
+	}
+
+	return []NamedNode{
+		{
+			name: dir,
+			node: OrderedDir{
+				Mode:    normalizeFileMode(0750 | mode),
+				ModTime: time.Now(),
+				Nodes:   getFileNodes(),
+			},
+		},
+	}
+}
+
+func setup(t *testing.T, namedNodes []NamedNode) *Restorer {
+	repo := repository.TestRepository(t)
+	sn, _ := saveOrderedSnapshot(t, repo, OrderedSnapshot{
+		nodes: namedNodes,
+	})
+	res := NewRestorer(repo, sn, false, nil)
+	return res
+}
+
+func saveDirOrdered(t testing.TB, repo restic.Repository, namedNodes []NamedNode, inode uint64) restic.ID {
```

Comment:
`saveDirOrdered`/`saveOrderedSnapshot` appears to massively duplicate code from `saveDir`. What has changed compared to that function? Why are those changes necessary? Why is it _not possible_ to extend `saveDir` as required?

Reply:
This was a byproduct of backporting these test cases from a more detailed ADS test with FileAttribute combinations. In case of ADS, we need to enforce an order in which ads related nodes are restored in order to cover certain scenarios. That ordering is not needed here. 
I have now cleaned up the code and reused saveDir and saveSnapshot.